File: src/lib/generators/transformationSql.ts
import type { MappingRow } from '@/types/mapping';
import { quoteColumn, qualifiedColumn } from '@/lib/sql/identifierQuoting';
import { resolveSourceReference, resolveTargetReference } from '@/lib/sql/sourceReference';
import type { TableTypeConfig } from '@/types/tableTypeConfig';
import {
  classifyTransformation,
  qualifyFieldReferences,
  type TransformationClassification,
} from '@/lib/generators/businessRuleHeuristics';

export interface FieldValidationSql {
  sql: string;
  isManualReview: boolean;
  classification: TransformationClassification;
}

interface KeyColumns {
  sourceCols: string[];
  targetCols: string[];
  usedFallbackKey: boolean;
}

/**
 * Some mapping docs cram more than one column into a single Source Column cell (e.g.
 * "column_1,att_1", when a transformation like concat(column_1,att_1) draws from two source
 * columns at once, sometimes across two source tables per the matching Source Table cell). Split
 * on comma/semicolon so each real column name is recoverable individually -- used only to widen
 * the known-field whitelist (§buildKnownFields), never to fabricate a single SQL identifier out of
 * the raw compound string.
 */
function splitCommaTokens(value: string): string[] {
  return value
    .split(/[,;]/)
    .map((t) => t.trim())
    .filter(Boolean);
}

/** True when a field name is a single clean identifier -- not a comma/semicolon-joined compound value that can't be used as a literal SQL column reference. */
function isSingleColumnName(value: string): boolean {
  return value.trim().length > 0 && !/[,;]/.test(value);
}

/** Picks the correlation key -- the declared primary key field(s), or (flagged with a comment) every mapped field as a best-effort fallback when no PK is declared. A compound (comma-joined) Source/Target Column value is never used as a key column -- there's no single real identifier to reference. */
function pickKeyColumns(tableRows: MappingRow[]): KeyColumns {
  const cleanRows = tableRows.filter((r) => isSingleColumnName(r.sourceField) && isSingleColumnName(r.targetField));
  const keyRows = cleanRows.filter((r) => r.isPrimaryKey);
  const effectiveKeyRows = keyRows.length > 0 ? keyRows : cleanRows;
  return {
    sourceCols: effectiveKeyRows.map((r) => r.sourceField),
    targetCols: effectiveKeyRows.map((r) => r.targetField),
    usedFallbackKey: keyRows.length === 0,
  };
}

/**
 * Builds the two separate, independently-runnable source and target queries for a field whose
 * transformation has already been resolved to a validated SQL expression -- shared by the
 * deterministic classifier's happy path below and by the browser-side AI Assist enrichment flow
 * (src/lib/llm/aiAssist.ts), which produces its own candidate expression from an LLM and, after
 * independently validating it against the same known-field/keyword whitelist, builds the exact
 * same query shape a deterministic match would have produced.
 */
export function buildSourceTargetQueries(
  row: MappingRow,
  tableRows: MappingRow[],
  typeConfig: TableTypeConfig,
  qualifiedExpr: string
): string {
  const sourceQualified = resolveSourceReference(typeConfig, tableRows, row.sourceSchema, row.sourceTable);
  const targetQualified = resolveTargetReference(typeConfig, row.targetSchema, row.targetTable);
  const { sourceCols, targetCols, usedFallbackKey } = pickKeyColumns(tableRows);
  const keyComment = usedFallbackKey
    ? `-- NOTE: no primary key flagged for this table; every mapped field is listed so rows can still be correlated by eye.\n`
    : '';

  const sourceKeySelect = sourceCols.map((f) => qualifiedColumn('s', f));
  const targetKeySelect = targetCols.map((f) => qualifiedColumn('t', f));
  const sourceOrderBy = sourceCols.length > 0 ? sourceKeySelect.join(', ') : qualifiedColumn('s', row.sourceField);
  const targetOrderBy = targetCols.length > 0 ? targetKeySelect.join(', ') : qualifiedColumn('t', row.targetField);

  const sourceSql = [
    `${keyComment}-- SOURCE query: derived_target_value is what the transformation should produce.`,
    `SELECT ${[...sourceKeySelect, `${qualifiedExpr} AS derived_target_value`].join(', ')}`,
    `FROM ${sourceQualified} s`,
    `ORDER BY ${sourceOrderBy};`,
  ].join('\n');

  const targetSql = [
    `-- TARGET query: actual_target_value is what was actually loaded -- compare rows by key against the source query above.`,
    `SELECT ${[...targetKeySelect, `${qualifiedColumn('t', row.targetField)} AS actual_target_value`].join(', ')}`,
    `FROM ${targetQualified} t`,
    `ORDER BY ${targetOrderBy};`,
  ].join('\n');

  return `${sourceSql}\n\n${targetSql}`;
}

/**
 * A transformation formula routinely references columns that aren't *this* row's own declared
 * source field: a sibling source column mapped elsewhere in the doc (e.g. "last_name" feeding a
 * "full_name" row whose own Source Field is "first_name"), or another already-computed target
 * column from this same table group (e.g. "discount_pct" derived from a target field
 * "order_total" that a different row already produces). Restricting knownFields to only this
 * row's own group's source fields rejected both patterns and forced them to MANUAL_REVIEW even
 * though the SQL was perfectly valid -- so the whitelist covers every source field mapped
 * anywhere in the workbook plus every target field within this table group. Exported so the
 * browser-side AI Assist flow validates an LLM's suggested expression against the identical set.
 */
export function buildKnownFields(tableRows: MappingRow[], allMappingRows: MappingRow[]): string[] {
  return [
    ...tableRows.map((r) => r.sourceField),
    ...tableRows.map((r) => r.targetField),
    ...allMappingRows.map((r) => r.sourceField),
  ]
    .filter(Boolean)
    .flatMap(splitCommaTokens);
}

export function buildFieldValidationSql(
  row: MappingRow,
  tableRows: MappingRow[],
  typeConfig: TableTypeConfig,
  allMappingRows: MappingRow[] = tableRows
): FieldValidationSql {
  const knownFields = buildKnownFields(tableRows, allMappingRows);
  const classification = classifyTransformation(row.transformation, knownFields);

  // Two separate, independently-runnable queries rather than one correlated JOIN -- the tester
  // runs each against its own side and compares the result sets by key, matching the same
  // source-query/target-query pattern already used for row-count reconciliation.
  if (classification.expression) {
    const qualifiedExpr = qualifyFieldReferences(classification.expression, knownFields, 's');
    const sql = buildSourceTargetQueries(row, tableRows, typeConfig, qualifiedExpr);
    return { sql, isManualReview: false, classification };
  }

  const sourceQualified = resolveSourceReference(typeConfig, tableRows, row.sourceSchema, row.sourceTable);
  const targetQualified = resolveTargetReference(typeConfig, row.targetSchema, row.targetTable);

  const sourceSql = [
    `-- SOURCE query (reference only -- the transformation below could not be auto-translated)`,
    `-- Raw transformation rule for ${row.targetField}: "${row.transformation}"`,
    `SELECT * FROM ${sourceQualified} s LIMIT 10;`,
  ].join('\n');

  const targetSql = [
    `-- TARGET query`,
    `-- MANUAL REVIEW REQUIRED: translate the rule above into a comparison against ${quoteColumn(row.targetField)} below.`,
    `SELECT * FROM ${targetQualified} t LIMIT 10;`,
  ].join('\n');

  const sql = `${sourceSql}\n\n${targetSql}`;

  return { sql, isManualReview: true, classification };
}

File: server/engine/transformation_sql.py
"""Direct port of src/lib/generators/transformationSql.ts."""

from __future__ import annotations

import re
from dataclasses import dataclass

from engine.business_rule_heuristics import (
    TransformationClassification,
    classify_transformation,
    qualify_field_references,
)
from engine.identifier_quoting import qualified_column, quote_column
from engine.models import MappingRow, TableTypeConfig
from engine.source_reference import resolve_source_reference, resolve_target_reference


@dataclass
class FieldValidationSql:
    sql: str
    is_manual_review: bool
    classification: TransformationClassification


def _split_comma_tokens(value: str) -> list[str]:
    return [t.strip() for t in re.split(r"[,;]", value) if t.strip()]


def _is_single_column_name(value: str) -> bool:
    return len(value.strip()) > 0 and not re.search(r"[,;]", value)


@dataclass
class _KeyColumns:
    source_cols: list[str]
    target_cols: list[str]
    used_fallback_key: bool


def _pick_key_columns(table_rows: list[MappingRow]) -> _KeyColumns:
    clean_rows = [r for r in table_rows if _is_single_column_name(r.source_field) and _is_single_column_name(r.target_field)]
    key_rows = [r for r in clean_rows if r.is_primary_key]
    effective_key_rows = key_rows if key_rows else clean_rows
    return _KeyColumns(
        source_cols=[r.source_field for r in effective_key_rows],
        target_cols=[r.target_field for r in effective_key_rows],
        used_fallback_key=len(key_rows) == 0,
    )


def build_known_fields(table_rows: list[MappingRow], all_mapping_rows: list[MappingRow]) -> list[str]:
    raw = (
        [r.source_field for r in table_rows]
        + [r.target_field for r in table_rows]
        + [r.source_field for r in all_mapping_rows]
    )
    result: list[str] = []
    for value in raw:
        if value:
            result.extend(_split_comma_tokens(value))
    return result


def build_source_target_queries(
    row: MappingRow, table_rows: list[MappingRow], type_config: TableTypeConfig, qualified_expr: str
) -> str:
    source_qualified = resolve_source_reference(type_config, table_rows, row.source_schema, row.source_table)
    target_qualified = resolve_target_reference(type_config, row.target_schema, row.target_table)
    keys = _pick_key_columns(table_rows)
    key_comment = (
        "-- NOTE: no primary key flagged for this table; every mapped field is listed so rows can still be correlated by eye.\n"
        if keys.used_fallback_key
        else ""
    )

    source_key_select = [qualified_column("s", f) for f in keys.source_cols]
    target_key_select = [qualified_column("t", f) for f in keys.target_cols]
    source_order_by = ", ".join(source_key_select) if keys.source_cols else qualified_column("s", row.source_field)
    target_order_by = ", ".join(target_key_select) if keys.target_cols else qualified_column("t", row.target_field)

    source_sql = "\n".join(
        [
            f"{key_comment}-- SOURCE query: derived_target_value is what the transformation should produce.",
            f"SELECT {', '.join([*source_key_select, f'{qualified_expr} AS derived_target_value'])}",
            f"FROM {source_qualified} s",
            f"ORDER BY {source_order_by};",
        ]
    )

    target_value_col = qualified_column("t", row.target_field)
    target_select = ", ".join([*target_key_select, f"{target_value_col} AS actual_target_value"])
    target_sql = "\n".join(
        [
            "-- TARGET query: actual_target_value is what was actually loaded -- compare rows by key against the source query above.",
            f"SELECT {target_select}",
            f"FROM {target_qualified} t",
            f"ORDER BY {target_order_by};",
        ]
    )

    return f"{source_sql}\n\n{target_sql}"


def build_field_validation_sql(
    row: MappingRow,
    table_rows: list[MappingRow],
    type_config: TableTypeConfig,
    all_mapping_rows: list[MappingRow] | None = None,
) -> FieldValidationSql:
    if all_mapping_rows is None:
        all_mapping_rows = table_rows
    known_fields = build_known_fields(table_rows, all_mapping_rows)
    classification = classify_transformation(row.transformation, known_fields)

    if classification.expression:
        qualified_expr = qualify_field_references(classification.expression, known_fields, "s")
        sql = build_source_target_queries(row, table_rows, type_config, qualified_expr)
        return FieldValidationSql(sql=sql, is_manual_review=False, classification=classification)

    source_qualified = resolve_source_reference(type_config, table_rows, row.source_schema, row.source_table)
    target_qualified = resolve_target_reference(type_config, row.target_schema, row.target_table)

    source_sql = "\n".join(
        [
            "-- SOURCE query (reference only -- the transformation below could not be auto-translated)",
            f'-- Raw transformation rule for {row.target_field}: "{row.transformation}"',
            f"SELECT * FROM {source_qualified} s LIMIT 10;",
        ]
    )
    target_sql = "\n".join(
        [
            "-- TARGET query",
            f"-- MANUAL REVIEW REQUIRED: translate the rule above into a comparison against {quote_column(row.target_field)} below.",
            f"SELECT * FROM {target_qualified} t LIMIT 10;",
        ]
    )

    sql = f"{source_sql}\n\n{target_sql}"
    return FieldValidationSql(sql=sql, is_manual_review=True, classification=classification)
